include/mpm_base.tcc

```diff
@@ -0,0 +1,628 @@
+bool mpm::MPMBase<Tdim>::initialise_materials() {
+  bool status = true;
+  try {
+    // Get materials properties
+    auto materials = io_->json_object("materials");
+
+    for (const auto material_props : materials) {
+      // Get material type
+      const std::string material_type =
+          material_props["type"].template get<std::string>();
+
+      // Get material id
+      auto material_id = material_props["id"].template get<unsigned>();
+
+      // Create a new material from JSON object
+      auto mat =
+          Factory<mpm::Material<Tdim>, unsigned, const Json&>::instance()
+              ->create(material_type, std::move(material_id), material_props);
+
+      // Add material to list
+      auto result = materials_.insert(std::make_pair(mat->id(), mat));
+
+      // If insert material failed
+      if (!result.second) {
+        status = false;
+        throw std::runtime_error(
+            "New material cannot be added, insertion failed");
+      }
+    }
+  } catch (std::exception& exception) {
+    console_->error("#{}: Reading materials: {}", __LINE__, exception.what());
+    status = false;
+  }
+  return status;
+}
+
+//! Apply nodal tractions
+template <unsigned Tdim>
+bool mpm::MPMBase<Tdim>::apply_nodal_tractions() {
+  bool status = true;
+  try {
+    // Read and assign nodes tractions
+    if (!io_->file_name("nodal_tractions").empty()) {
+      // Get mesh properties
+      auto mesh_props = io_->json_object("mesh");
+      // Get Mesh reader from JSON object
+      const std::string reader =
+          mesh_props["mesh_reader"].template get<std::string>();
+      // Create a mesh reader
+      auto node_reader =
+          Factory<mpm::ReadMesh<Tdim>>::instance()->create(reader);
+
+      bool nodal_tractions =
+          mesh_->assign_nodal_tractions(node_reader->read_particles_tractions(
+              io_->file_name("nodal_tractions")));
+      if (!nodal_tractions)
+        throw std::runtime_error("Nodal tractions are not properly assigned");
+    } else
+      nodal_tractions_ = false;
+  } catch (std::exception& exception) {
+    console_->error("#{}: Nodal traction: {}", __LINE__, exception.what());
+    status = false;
+    nodal_tractions_ = false;
+  }
+  return status;
+}
+
+//! Apply properties to particles sets (e.g: material, remove step)
+template <unsigned Tdim>
+bool mpm::MPMBase<Tdim>::apply_properties_to_particles_sets() {
+  bool status = false;
+  // Assign material to particle sets
+  try {
+    // Get particle properties
+    auto particle_props = io_->json_object("particle");
+    // Get particle sets properties
+    auto particle_sets = particle_props["particle_sets"];
+    // Iterate over each particle sets
+    for (const auto& psets : particle_sets) {
+      // Assign material to particle sets
+      // Get set material from list of materials
+      auto set_material = materials_.at(psets["material_id"]);
+      // Get sets ids
+      std::vector<unsigned> sids = psets["set_id"];
```

Comment:
Let's have each particle set defined separately in the input.json file so it does not complicate when more properties are defined.
i.e. `set_id` is one  value, not a vector.